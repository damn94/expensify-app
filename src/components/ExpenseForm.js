//single date picker
import {SingleDatePicker} from 'react-dates'
import moment from 'moment';
import React from "react";

class ExpenseForm extends React.Component {
//to change props constructor function
constructor(props){
	super(props);
	this.state = {
 		description:props.expense ? props.expense.description: '',
 		note: props.expense ? props.expense.note : '',
 		amount:props.expense ? (props.expense.amount /100).toString() : '',
 		createdAt: props.expense ? moment(props.expense.createdAt): moment(),
 		calendarFocus: false,
 		error: undefined
	}
}



//on description change function
onDescriptionChange = (e) => {
	const description = e.target.value
	this.setState(()=>({description}))
}
	

//on note change function
onNoteChange = (e) => {
	const note = e.target.value
	this.setState(()=>({note}))
}

//on amount change's function
onAmountChange = (e) => {
const amount = e.target.value
if (!amount || amount.match(/^\d{1,}(\.\d{0,2})?$/)){
		this.setState(()=>({amount}))
	}
}

//on date change for component SingleDatePicker
onDateChange = (createdAt) =>{
	if (createdAt) {
		this.setState(()=>({createdAt}))
	}
}


//ON FOCUS CHANGE
onFocusChange = ({focused}) => {
	this.setState(()=>({calendarFocus: focused}))
}


//ON SUBMIT CHANGE
onSubmitChange = (e) => {
	e.preventDefault()
	if (!this.state.amount || !this.state.description){
		const error = 'Please,provide description and  amount.'
		this.setState(()=>({error}))
	}else{
		this.setState(()=> ({error: ''}))
			//function send data to AddExpensePage Component
			this.props.onSubmit({
			description: this.state.description,
			createdAt: this.state.createdAt.valueOf(),
			note: this.state.note,
			amount: parseFloat(this.state.amount, 10) * 100
		})
	}

}

render(){
	return (
		<div>

			{this.state.error && <h4>{this.state.error}</h4>}
			{/*keep track input field change*/}

			{/*description filed*/}
		<form onSubmit={this.onSubmitChange}>
			<input 
			type='text' 
			placeholder='description' 
			value={this.state.description} 
			autoFocus
			onChange={this.onDescriptionChange}
			/>
			
			{/*amount field*/}
			<input	 
			type='text' 
			value={this.state.amount} 
			placeholder='amount' 
			onChange={this.onAmountChange}
			/>

			<SingleDatePicker 
			date={this.state.createdAt}
			onDateChange={this.onDateChange}
			focused={this.state.calendarFocus}
			onFocusChange={this.onFocusChange}
			numberOfMonths={1}
			isOutsideRange={()=> false}
			/>
			
			{/*note field*/}
			<textarea 
			placeholder='Add your expense note here! (optional)'
			value={this.state.note}
			onChange={this.onNoteChange}
			/>

			<button>Add Expense</button>
		</form>
		</div>
		)
	}
}


export default ExpenseForm;