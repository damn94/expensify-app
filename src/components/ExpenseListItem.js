import {Link} from 'react-router-dom'
import  React from 'react';

export const ExpenseListItem = ({description , amount, createdAt,id }) => (
	<div>
		<h3>
			<Link to={`/edit/${id}`}>
				{description}
			</Link>
		</h3>

		<p>{amount} -- {createdAt}</p>
	</div>
)


export default ExpenseListItem;