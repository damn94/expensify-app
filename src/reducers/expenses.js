

//defaulvalue state expense reducer
const expenseReducerDefaultState = [] ;

//EXPENSE REDUCER
export default (state=expenseReducerDefaultState,action) =>{
switch(action.type){
	//CASE ADD_EXPENSE
	case 'ADD_EXPENSE':
		return [
			...state,
			action.expense
		];
	//CASE REMOVE_EXPENSE
	case 'REMOVE_EXPENSE':
	 return state.filter((expense)=> expense.id !== action.id);

	 //CASE EDIT_EXPENSE
	case 'EDIT_EXPENSE':
		return state.map((expense) =>{
			if(expense.id === action.id){
				return {...expense,...action.updates}
			}else{
				return expense;
			}
		}) 

	 //DEFAULT
	default:
		return state
	};
}