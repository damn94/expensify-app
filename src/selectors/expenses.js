import moment from 'moment';

//create a function that can filter store
export default (expenses, {text,sortBy,endDate,startDate}) =>{

return expenses.filter((expense)=>{
	const createdAtMoment =	moment(expense.createdAt)
	const startDateMatch = startDate ? startDate.isSameOrBefore(createdAtMoment,'day'):true;
	const endDateMatch = endDate ? endDate.isSameOrAfter(createdAtMoment,'day') : true;
	const textMatch = expense.description.toLowerCase().includes(text.toLowerCase());

	return startDateMatch && endDateMatch && textMatch
	}).sort((a,b)=>{
//SORTING THE ARRAY
		if (sortBy === 'date'){
			//SORTING BY DATE

			return a.createdAt > b.createdAt ? -1 : 1
		}else if (sortBy === 'amount') {
			
			//SORTING BY AMOUNT
			return a.amount > b.amount ? -1 : 1
		}
	})
}
