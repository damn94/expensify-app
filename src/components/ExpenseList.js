import { connect } from 'react-redux';
import React from 'react';
import ExpenseListItem from './ExpenseListItem'
import selectorsExpense from '../selectors/expenses'

export const ExpenseList = (props) =>(
	<div>
	{

	props.expenses.length === 0 ?(
			<h1>No expenses</h1>
		):(
		props.expenses.map((expense)=>{
			return <ExpenseListItem key={expense.id}{...expense}/>
		})

		)
	}
	</div>
)
const mapStateToProps = (state) =>{
	return {
		expenses: selectorsExpense(state.expenses,state.filters)
	}
}

export default connect(mapStateToProps)(ExpenseList);