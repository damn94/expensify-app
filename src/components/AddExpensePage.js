import {addExpense} from '../actions/expenses'
import ExpenseForm from './ExpenseForm';
import {connect} from 'react-redux';
import React from 'react';


//convert component ADDEXPENSEDASHBORDPAGE to class  
export class AddExpensePage extends React.Component{

addExpense = (expense)=> {
		this.props.addExpense(expense)
		this.props.history.push('/')
	} 

render(){

return(
		<div>
			<ExpenseForm onSubmit={this.addExpense}/>
		</div>
	)}
}





// //COMPONENT ADDEXPENSEDASHBORDPAGE
// const AddExpensePage = (props) => (
// 	<div>
// 		<ExpenseForm onSubmit={
// 		//grap the input from the browser => store
// 		//add Expense
// 			(data)=> {
// 				//props.dispatch(addExpense(data))
// 				props.onSubmit(data)
// 				props.history.push('/')
// 			}
// 		}/>
// 	</div>
// )

//the second argument for props
const mapDispatchToProps = (dispatch) => ({
	addExpense: (expense) => dispatch(addExpense(expense))
})

export default connect(undefined,mapDispatchToProps)(AddExpensePage);