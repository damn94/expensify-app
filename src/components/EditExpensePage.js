import {editExpense,removeExpense} from '../actions/expenses'
import ExpenseForm from './ExpenseForm';
import {connect} from 'react-redux';
import React from 'react';


//refactoring component to class
export class EditExpensePage extends React.Component {

removeExpense =()=> {

	this.props.removeExpense({id: this.props.expense.id})
	this.props.history.push('/')
}
editExpense = (expense)=>{
	this.props.editExpense(this.props.expense.id,expense)
	this.props.history.push('/')
}

render(){
	return(
		<div>
			<ExpenseForm expense={this.props.expense} onSubmit={this.editExpense}/>

			{/*remove button*/}
			<button onClick={this.removeExpense}>Remove</button>
			
		</div>	
	)		
	}
}




//COMPONENT EDITEXPENSEDASHBORDPAGE
// const EditExpensePage = (props) =>{
// return(
// 		<div>
// 			<ExpenseForm {...props} onSubmit={(expense)=>{
// 				props.dispatch(editExpense(props.match.params.id,expense))
// 				props.history.push('/')
// 			}}/>

// 			{/*remove button*/}
// 			<button onClick={()=>{
// 				props.dispatch(removeExpense({id: props.expense.id}))
// 				props.history.push('/')
// 			}}>Remove</button>
			
// 		</div>
// 	)
// }

const mapDispatchToProps = (dispatch, props) =>( {
	editExpense: (id,expense) => dispatch(removeExpense(id, expense)),
	removeExpense: (data)=> dispatch(removeExpense(data))
})


const mapStateToProps = (state,props) =>{
	return {
		expense: state.expenses.find((expense)=>expense.id === props.match.params.id
	)}
}

export default connect(mapStateToProps,mapDispatchToProps)(EditExpensePage);