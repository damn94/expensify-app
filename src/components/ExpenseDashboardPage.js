import React from 'react';
import ExpenseList	from './ExpenseList'
import ExpenseListFilters from './ExpenseListFilters'
//COMPONENT EXPENSEDASHBORDPAGE
const ExpenseDashboardPage = () => (
	<div>
		<ExpenseListFilters />
		<ExpenseList />
	</div>
)



export default ExpenseDashboardPage;


