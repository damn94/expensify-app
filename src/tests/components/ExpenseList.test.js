import { ExpenseList } from '../../components/ExpenseList'
import expenses from '../fixtures/expenses'
import {shallow} from 'enzyme'
import React from 'react';


//test the not connected version of  ExpenseList
test('Should render ExpenseList with expenses',()=>{

	const wrapper = shallow(<ExpenseList expenses={expenses}/>)
	
	expect(wrapper).toMatchSnapshot()
})

//case expenses.length = 0

test('Should render ExpenseList with empty message',()=>{
	const wrapper = shallow(<ExpenseList expenses={[]} />)
	expect(wrapper).toMatchSnapshot()

})