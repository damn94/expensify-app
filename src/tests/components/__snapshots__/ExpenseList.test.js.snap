// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should render ExpenseList with empty message 1`] = `
<div>
  <h1>
    No expenses
  </h1>
</div>
`;

exports[`Should render ExpenseList with expenses 1`] = `
<div>
  <ExpenseListItem
    amount={73}
    createdAt={345600000}
    description="Tea"
    id="1"
    key="1"
    note=""
  />
  <ExpenseListItem
    amount={0.73}
    createdAt={0}
    description="Glass"
    id="2"
    key="2"
    note=""
  />
  <ExpenseListItem
    amount={297}
    createdAt={-345600000}
    description="Sugar"
    id="3"
    key="3"
    note=""
  />
</div>
`;
